### PR TITLE
[#706] Help request Form

### LINKED ISSUE #706: Help request Form / Mail
IMPORTANT!!! Expert email isn't known, for now, it will have value xyz@mail.pl

Based on:
https://g3v8he.axshare.com/#g=1&p=scen1

Modal 1:
https://invis.io/S9RJNE8WXUB#/431670848_ZK_-_Help_Request_-_1-2

Modal 2:
https://invis.io/S9RJNE8WXUB#/431670849_ZK_-_Help_Request_-_2-2

- New functionality (modal) Wsparcie eksperckie on /map in zk options dropdown
  - Send help request to an expert email by application API
  - Send confirmation email to a user via email of the request submission
  - Show success notification on correct apply
  - Confirm with second modal

Modal will contain
- Description where an answear will be send (on user email)
- Rodzaj wsparcia select input with options: wsparcie zdalne, wsparcie wyjazdowe
- Opis pożądanego zakresu wsparcia as tekstarea with message

### PR DESCRIPTION
**Pull request by danielkryska**
IMPORTANT!!! Expert email is not known, for now, it will have value xyz@mail.pl

Modal 1:
https://invis.io/S9RJNE8WXUB#/431670848_ZK_-_Help_Request_-_1-2

Modal 2:
https://invis.io/S9RJNE8WXUB#/431670849_ZK_-_Help_Request_-_2-2

- New functionality (modal) Wsparcie eksperckie on /map in zk options dropdown (button with lightning)
  - Confirm sending after click `Wyślij`
  - Send help request to an expert email by application API
  - Send confirmation email to a user via email of the request submission
  - Show success notification on correct apply

Modal will contain
- Description where an answear will be send (on user email)
- Rodzaj wsparcia select input with options: wsparcie zdalne, wsparcie wyjazdowe
- Opis pożądanego zakresu wsparcia as tekstarea with message

Closes #706

### CODE REVIEW

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/ExpertHelpController.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+@RestController
+@RequestMapping(path = API_PREFIX_V1, produces = APPLICATION_JSON_VALUE)
+@RequiredArgsConstructor
+@Tag(name = "expertHelp", description = "Send issue to helpdesk")
+public class ExpertHelpController {
+    private final ApplicationEventPublisher eventPublisher;
+
+    @Operation(summary = "Send invitation email to user")
+    @ApiResponses({
+            @ApiResponse(responseCode = "200", description = "Invitation was send"),
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/controller/ExpertHelpController.java:34]: Correct the description

File: s4e-backend/src/main/java/pl/cyfronet/s4e/config/SecurityConfig.java [hunk 2]
```diff
@@ -84,6 +84,8 @@ private void authorizeRequests(ExpressionUrlAuthorizationConfigurer<HttpSecurity
                 .mvcMatchers(POST, prefix("/invitations/{token}/confirm")).authenticated()
                 .mvcMatchers(PUT, prefix("/invitations/{token}/reject")).permitAll()
 
+                .mvcMatchers(POST, prefix("/expert-help")).authenticated()
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/config/SecurityConfig.java:87]: Access should be limited to ZK members if I understand correctly

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/request/HelpType.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package pl.cyfronet.s4e.controller.request;
+
+public enum HelpType {
+    /**
+     * Trough internet
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/controller/request/HelpType.java:5]: Trough -> Through

File: s4e-backend/src/main/java/pl/cyfronet/s4e/listener/ExpertHelpListener.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import org.thymeleaf.context.Context;
+import pl.cyfronet.s4e.event.OnSendHelpRequestEvent;
+import pl.cyfronet.s4e.service.InstitutionService;
+import pl.cyfronet.s4e.service.MailService;
+import pl.cyfronet.s4e.util.MailHelper;
+
+@Component
+@RequiredArgsConstructor
+public class ExpertHelpListener {
+    public final static String expertMail = "xyz@mail.pl";
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/listener/ExpertHelpListener.java:19]: I'd externalize this somewhere. Either to system property or the property DB table

File: s4e-backend/src/main/java/pl/cyfronet/s4e/listener/ExpertHelpListener.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    @EventListener
+    @Transactional(readOnly = true)
+    public void handle(OnSendHelpRequestEvent event) {
+        sendEmailToExpertBy(event);
+        sendConfirmationEmailToUserBy(event);
+    }
+
+    private void sendConfirmationEmailToUserBy(OnSendHelpRequestEvent event) {
+        String subject = messageSource.getMessage(
+                "email.email.expert-help-sending-confirmation.subject.subject",
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/listener/ExpertHelpListener.java:37]: You have duplicated email and subject in the property name.
Also, the property is undefined.

**Review by jswk — CHANGES_REQUESTED** [review]
There are several things I'd see changed, otherwise seems fine.

Also, the modal doesn't disappear after sending the request. I'd expect it to disappear.

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/ExpertHelpController.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import static pl.cyfronet.s4e.Constants.API_PREFIX_V1;
+
+@RestController
+@RequestMapping(path = API_PREFIX_V1, produces = APPLICATION_JSON_VALUE)
+@RequiredArgsConstructor
+@Tag(name = "expertHelp", description = "Send issue to helpdesk")
+public class ExpertHelpController {
+    private final ApplicationEventPublisher eventPublisher;
+
+    @Operation(summary = "Send invitation email to user")
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/controller/ExpertHelpController.java:32]: You missed this.

File: s4e-web/src/app/views/map-view/zk/expert-help-modal/expert-help-modal.component.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    if (await this.modalService.confirm(
+      'Wsparcie eksperckie',
+      'Czy na pewno chcesz wysłać prośbę o wsparcie zdalne?'
+    )) {
+      this._expertHelpService.sendHelpRequest$(this.form.value)
+      .subscribe(() => {
+        this._notificationService.addGeneral({
+          content: 'Prośba o wsparcie experckie została wysłana',
```
Comment (jswk) [inline s4e-web/src/app/views/map-view/zk/expert-help-modal/expert-help-modal.component.ts:66]: experckie -> eksperckie

**Review by jswk — CHANGES_REQUESTED** [review]
Thanks for adding correct permission requirements and other fixes. Apart from the two comments, add the sample env value to `backend-production.env.template`, as we discussed. Something along the lines: `EXPERT_HELP_MAIL=<email to expert>`.
`

**Review by jswk — APPROVED** [review]

**Comment by jswk** [comment]
I took this over from @danielkryska, changing the expert email storage to a DB property, instead of a runtime property, as sensibly suggested by @wziajka (my bad for suggesting a property in the first place).

**Review by abacz — APPROVED** [review]

**Pull request merged**